### PR TITLE
[FLINK-33262][table-api] Extend source provider interfaces with the new parallelism provider interface

### PR DESCRIPTION
**Pull request by X-czh**
## What is the purpose of the change

Extend source provider interfaces with the new parallelism provider interface. It includes all the public APi changes.

## Brief change log

- Extend SourceProvider with ParallelismProvider
- Extend InputFormatProvider with ParallelismProvider
- Introduce SOURCE_PARALLELISM option

Extension for the SourceFunctionProvider is omitted due to its deprecation.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by X-czh** [comment]
@libenchao Could you help review it when you have time? Many thanks in advance.

**Review by libenchao — COMMENTED** [review]
This the FLIP, you declared to support 5 source providers, however you only touched 2 of them, is there any consideration behind?

**Comment by X-czh** [comment]
@libenchao Thanks for the review. For `SourceFunctionProvider`, since it has already been deprecated now, I think there's no need to extend its interface, WDYT? For the other two providers involving interoperation with DataStream API,  sorry that I missed them, I'll extend them as well

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/InputFormatProvider.java [hunk 1]
```diff
@@ -44,6 +50,29 @@ public boolean isBounded() {
         };
     }
 
+    /** Helper method for creating a static provider with a provided source parallelism. */
+    static InputFormatProvider of(
```
[the remark this answers is not in the document]
Comment (X-czh) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/InputFormatProvider.java:54]: Good idea, I'll do the same for `SourceProvider` as well

**Comment by libenchao** [comment]
> @libenchao For `SourceFunctionProvider`, since it has already been deprecated now, I think there's no need to extend its interface, WDYT? 

@X-czh I agree we can leave deprecated interfaces out. However, I think it would still be nice to do it for `SourceFunctionProvider` for the following reasons:
* `SourceFunction` is still used in some third party connectors
* it's trivial to support it
* `SinkFunctionProvider` has supported it

**Comment by X-czh** [comment]
@libenchao Got it. Thanks for the detailed clarification! I'll extend the interface for `SourceFunctionProvider` as well.

**Comment by X-czh** [comment]
I've pushed two new commits: 
- one addresses Benchao's comments on extending `SourceFunctionProvider`, `DataStreamScanProvider`, and `TransformationScanProvider` as well as optimizing the current impl by removing redundant code.
- the other updates the comments of `ParallelismProvider` to reflect the new changes brought by the FLIP.

**Review by libenchao — APPROVED** [review]
LGTM, @X-czh thanks for your contribution!

**Pull request closed without merging**